scipy/integrate/_cubature.py

```diff
@@ -0,0 +1,333 @@
+import heapq
+import itertools
+
+from dataclasses import dataclass
+from scipy._lib._util import MapWrapper
+
+import numpy as np
+
+from scipy.integrate._rules import (
+    ProductNestedFixed, NestedFixedRule,
+    GaussKronrodQuadrature, NewtonCotesQuadrature,
+    GenzMalikCubature,
+)
+
+
+@dataclass
+class CubatureRegion:
+    estimate: np.ndarray
+    error: np.ndarray
+    a: np.ndarray
+    b: np.ndarray
+
+    def __lt__(self, other):
+        # Consider regions with higher error estimates as being "less than" regions with
+        # lower order estimates, so that regions with high error estimates are placed at
+        # the top of the heap.
+        return _max_norm(self.error) > _max_norm(other.error)
+
+
+@dataclass
+class CubatureResult:
+    estimate: np.ndarray
+    error: np.ndarray
+    success: bool
+    status: str
+    regions: list[CubatureRegion]
+    subdivisions: int
+    atol: float
+    rtol: float
+
+
+def cubature(f, a, b, rule="gk21", rtol=1e-05, atol=1e-08, max_subdivisions=10000,
+             args=(), kwargs=None, workers=1):
+    r"""
+    Adaptive cubature of multidimensional array-valued function.
+
+    Given an arbitrary integration rule, this function returns an estimate of the
+    integral to the requested tolerance over the region defined by the arrays `a` and
+    `b` specifying the corners of a hypercube.
+
+    Convergence is not guaranteed for all integrals.
+
+    Parameters
+    ----------
+    f : callable
+        Function to integrate. `f` must have the signature::
+            f(x : ndarray, \*args, \*\*kwargs) -> ndarray
+
+        `f` should accept arrays ``x`` of shape::
+            (npoints, ndim)
+
+        and output arrays of shape::
+            (npoints, output_dim_1, ..., output_dim_n)
+
+        In this case, `cub` will return arrays of shape::
+            (output_dim_1, ..., output_dim_n)
+    a, b : array_like or float
+        Lower and upper limits of integration as rank-1 arrays specifying the left and
+        right endpoints of the intervals being integrated over. If a float is passed,
+        these will be converted to singleton arrays. Infinite limits are currently not
+        supported.
+    rule : str, optional
+        Rule used to estimate the integral. If passing a string, the options are
+        "gauss-kronrod" (21 node), "newton-cotes" (5 node) or "genz-malik" (degree 7).
```

Comment:
The main motivation for offering it is so that it supports the same values as `quad_vec`'s `quadrature` argument but if you don't think that's a concern, I'm happy to remove it.